### PR TITLE
Integration with 3Dconnexion input devices

### PR DESCRIPTION
**Pull request by pskowronskiTDx**
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

The contribution was created in whole or in part by me and I
have the right to submit it under the open source license
indicated in the file.

I understand and agree that this project and the contribution
are public and that a record of the contribution (including all
personal information I submit with it, including my sign-off) is
maintained indefinitely and may be redistributed consistent with
this project or the open source license(s) involved.

### CODE REVIEW

File: avogadro/rendering/camera.cpp
Comment (ghutchis) [inline avogadro/rendering/camera.cpp:1]: Please don't commit files with whitespace changes, unless these come from running `clang-format`

File: avogadro/rendering/camera.h [hunk 1]
```diff
@@ -120,7 +120,21 @@ class AVOGADRORENDERING_EXPORT Camera
    * @param zFar is the distance from the viewer to the far clipping plane.
    */
   void calculatePerspective(float fieldOfView, float zNear, float zFar);
-
+#ifdef _3DCONNEXION
```
Comment (ghutchis) [inline avogadro/rendering/camera.h:123]: This seems like a perfectly reasonable API addition. In that case, please do not `#ifdef` the code and put the implementation properly in `camera.cpp`.

File: avogadro/rendering/geometryvisitor.cpp [hunk 2]
```diff
@@ -9,6 +9,9 @@
 #include "curvegeometry.h"
 #include "linestripgeometry.h"
 #include "spheregeometry.h"
+#ifdef _3DCONNEXION
+#include "cylindergeometry.h"
```
Comment (ghutchis) [inline avogadro/rendering/geometryvisitor.cpp:13]: Why is `cylindergeometry.h` needed?

File: avogadro/rendering/geometryvisitor.cpp [hunk 3]
```diff
@@ -27,6 +30,9 @@ void GeometryVisitor::visit(Drawable&)
 
 void GeometryVisitor::visit(SphereGeometry& geometry)
```
Comment (ghutchis) [inline avogadro/rendering/geometryvisitor.cpp:31]: What are you trying to do here?

File: avogadro/rendering/geometryvisitor.h [hunk 4]
```diff
@@ -61,7 +71,34 @@ class GeometryVisitor : public Visitor
    * Get the radius of the scene.
    */
   float radius();
+#ifdef _3DCONNEXION
```
Comment (ghutchis) [inline avogadro/rendering/geometryvisitor.h:74]: Again, this seems like a perfectly reasonable API addition without the `#ifdef` - and thus the implementation should be in the relevant `.cpp` file.

File: avogadro/rendering/geometryvisitor.h [hunk 5]
```diff
@@ -41,7 +47,11 @@ class GeometryVisitor : public Visitor
   void visit(SphereGeometry&) override;
   void visit(AmbientOcclusionSphereGeometry&) override;
   void visit(CurveGeometry&) override;
+#ifdef _3DCONNEXION
+  void visit(CylinderGeometry&) override;
```
Comment (ghutchis) [inline avogadro/rendering/geometryvisitor.h:51]: If you want to replace this `visit()` method, you should do that (and whatever else you want to do in the implementation rather than an `ifdef` in the header.

File: avogadro/rendering/geometryvisitor.h [hunk 6]
```diff
@@ -25,6 +25,12 @@ namespace Rendering {
  * notably the center and radius of the bounding sphere.
  */
 
+#ifdef _3DCONNEXION
```
Comment (ghutchis) [inline avogadro/rendering/geometryvisitor.h:28]: Why are these needed here?

File: avogadro/rendering/geometryvisitor.h [hunk 7]
```diff
@@ -61,7 +71,34 @@ class GeometryVisitor : public Visitor
...
+   * @param maxY [out] maximum Y coordinate of the box diagonal
+   * @param maxZ [out] maximum Z coordinate of the box diagonal
+   * @param flags [in] flags informing which atoms will be included
+   * in the bounding box.
+   */
+  void boundingBox(double& minX, double& minY, double& minZ, double& maxX,
+                   double& maxY, double& maxZ,
+                   const std::vector<bool>& flags) const;
 
+  /**
```
Comment (ghutchis) [inline avogadro/rendering/geometryvisitor.h:91]: Again. This seems like a reasonable API addition. Add it with no `#ifdef` and then the appropriate implementation.

File: avogadro/rendering/geometryvisitor.h [hunk 8]
```diff
@@ -74,6 +111,11 @@ class GeometryVisitor : public Visitor
 
   std::vector<Vector3f> m_centers;
   std::vector<float> m_radii;
+#ifdef _3DCONNEXION
```
Comment (ghutchis) [inline avogadro/rendering/geometryvisitor.h:114]: Why are you saving all of these? This seems like a really bad idea when someone has 10^5 or 10^6 atoms and bonds.

File: avogadro/rendering/glrenderer.cpp [hunk 9]
```diff
@@ -27,8 +27,25 @@ using Core::Array;
...
   : m_valid(false), m_textRenderStrategy(nullptr), m_center(Vector3f::Zero()),
     m_radius(20.0)
+#ifdef _3DCONNEXION
+    ,
+    m_drawIcon(false), m_iconData(nullptr), m_iconWidth(0u), m_iconHeight(0u),
+    m_iconPosition(Eigen::Vector3f::Zero())
+#endif
 {
   m_overlayCamera.setIdentity();
+#ifdef _3DCONNEXION
```
Comment (ghutchis) [inline avogadro/rendering/glrenderer.cpp:37]: I would be willing to see how this works in general. It doesn't seem specific to your code.

File: avogadro/rendering/glrenderer.cpp [hunk 10]
```diff
@@ -169,16 +204,43 @@ void GLRenderer::setTextRenderStrategy(TextRenderStrategy* tren)
 void GLRenderer::applyProjection()
 {
   float distance = m_camera.distance(m_center);
+#ifdef _3DCONNEXION
```
Comment (ghutchis) [inline avogadro/rendering/glrenderer.cpp:207]: I think I would rather test this as a general change rather than a bunch of `#ifdef` here.

File: avogadro/rendering/scene.h [hunk 11]
```diff
@@ -130,11 +130,39 @@ class AVOGADRORENDERING_EXPORT Scene
 
   /** Clear the scene of all elements. */
   void clear();
-
+#ifdef _3DCONNEXION
```
Comment (ghutchis) [inline avogadro/rendering/scene.h:133]: Again, if you want to add API, just add API and associated implementation.

File: avogadro/rendering/tdxextensions.cpp
Comment (ghutchis) [inline avogadro/rendering/tdxextensions.cpp:1]: This file should not exist. Methods belong in their appropriate `.cpp` files so people can find them. Thanks.

**Review by ghutchis — CHANGES_REQUESTED** [review]
In general, these seem reasonable. I'd much, much rather see new API methods and implementations in their associated files than a `tdxextensions.cpp` with some miscellaneous methods that belong elsewhere.

The main `ifdef` pieces I'd want to see are relative to drawing icons, etc.

**Comment by ghutchis** [comment]
@pskowronskiTDx - any comments / concerns about the requested changes? Happy to discuss over e-mail if that's easier. Thanks!

File: avogadro/rendering/camera.cpp
Comment (pskowronskiTDx) [reply avogadro/rendering/camera.cpp:1]: Sure, I believe I have added them accidentaly.

File: avogadro/rendering/geometryvisitor.cpp (hunk 2 above)
Comment (pskowronskiTDx) [reply avogadro/rendering/geometryvisitor.cpp:13]: Before I moved part of the implementation to tdxextensions.cpp there were an overriden "visit" method for cylinder geometry here. I forgot to remove this include.

File: avogadro/rendering/geometryvisitor.h (hunk 6 above)
Comment (pskowronskiTDx) [reply avogadro/rendering/geometryvisitor.h:28]: They were added to cache scene geometry for hit testing and it's bounding box calculation.

File: avogadro/rendering/geometryvisitor.cpp (hunk 3 above)
Comment (pskowronskiTDx) [reply avogadro/rendering/geometryvisitor.cpp:31]: Cache the scene geometry.

File: avogadro/rendering/geometryvisitor.h (hunk 8 above)
Comment (pskowronskiTDx) [reply avogadro/rendering/geometryvisitor.h:114]: I agree, the hit testing part introduced by me is redundant and it can be memory consuming as you noted. I have reworked it today.

File: avogadro/rendering/glrenderer.cpp (hunk 9 above)
Comment (pskowronskiTDx) [reply avogadro/rendering/glrenderer.cpp:37]: Ok, I will remove mentioned #ifdef's from GLRenderer making my changes to this class general.

File: avogadro/rendering/geometryvisitor.h (hunk 6 above)
Comment (pskowronskiTDx) [reply avogadro/rendering/geometryvisitor.h:28]: It seems to be an another leftover. Removed.

**Comment by pskowronskiTDx** [comment]
@ghutchis - Here is a summary of changes I did according to your review. I hope you wil be happy with them,

- The definitions of new methods have been putted back to relevant source files, tdxextensions.cpp has been removed.
- The #ifdef conditions got removed for Scene, Camera and GLRenderer (except for part responsible for drawing the 3Dx icon) classes.
- Saving the geometry got reduced to spheres. The point of doing so is to access the current collection of the spheres visible on scene. It is needed to calculate bounding box of the whole molecule or it's part consisting of selected atoms. I couldn't come up with more efficient solution to this problem so far. Maybe there is another approach you can suggest to follow?
- Sources were cleaned-up of whitespace changes and unnecessary includes.

**Comment by ghutchis** [comment]
I'm traveling today, but will take a look. @cryos - do you have a moment to look as well?

**Comment by pskowronskiTDx** [comment]
Hi Geoff, do you managed to take a look on the changes I have proposed to both PRs? Thanks.

File: avogadro/rendering/camera.cpp [hunk 12]
```diff
@@ -23,7 +23,7 @@ Camera::Camera()
 Camera::Camera(const Camera& o)
   : m_width(o.m_width), m_height(o.m_height),
     m_projectionType(o.m_projectionType),
-    m_orthographicScale(o.m_orthographicScale), m_data(new EigenData(*o.m_data))
+    m_orthographicScale(o.m_orthographicScale), m_data(new EigenData(*o.m_data))
```
Comment (cryos) [inline avogadro/rendering/camera.cpp:26]: This just adds a trailing space and should be fixed.

File: avogadro/rendering/CMakeLists.txt [hunk 13]
```diff
@@ -144,3 +145,7 @@ target_link_libraries(Rendering
     Avogadro::Core # Needed for an emum in residue.h that we should move, maybe others.
     GLEW::GLEW
     OpenGL::GL)
+
+if(USE_3DCONNEXION AND (WIN32 OR APPLE))
+  target_compile_definitions(Rendering PUBLIC _3DCONNEXION)
```
Comment (cryos) [inline avogadro/rendering/CMakeLists.txt:150]: Does this need to be public?

File: avogadro/rendering/glrenderer.h [hunk 14]
```diff
@@ -136,7 +152,7 @@ class AVOGADRORENDERING_EXPORT GLRenderer
   Scene m_scene;
   TextRenderStrategy* m_textRenderStrategy;
   SolidPipeline m_solidPipeline;
-
+
```
Comment (cryos) [inline avogadro/rendering/glrenderer.h:155]: Just trailing space, please remove.

**Comment by ghutchis** [comment]
> * Saving the geometry got reduced to spheres. The point of doing so is to access the current collection of the spheres visible on scene. It is needed to calculate bounding box of the whole molecule or it's part consisting of selected atoms. I couldn't come up with more efficient solution to this problem so far. Maybe there is another approach you can suggest to follow?

Can't you calculate the bounding box (or selected atom bounding) in `core/molecule.*` or `qtgui/molecule.*` instead? There are a few places that use similar code.

**Comment by ghutchis** [comment]
Please just run `clang-format` on your modified files, thanks.

File: avogadro/rendering/camera.h [hunk 15]
```diff
@@ -121,6 +121,20 @@ class AVOGADRORENDERING_EXPORT Camera
    */
   void calculatePerspective(float fieldOfView, float zNear, float zFar);
 
+  /**
+   * << API Extension for TDX >>
```
Comment (ghutchis) [inline avogadro/rendering/camera.h:125]: Please omit this. You may have added it, but there's plenty of reason to use this in other code.

File: avogadro/rendering/geometryvisitor.cpp [hunk 16]
```diff
@@ -187,4 +189,55 @@ void GeometryVisitor::average()
   }
 }
 
+void GeometryVisitor::boundingBox(double& minX, double& minY, double& minZ,
```
Comment (ghutchis) [inline avogadro/rendering/geometryvisitor.cpp:192]: As I mentioned, it seems easier to put this into `core/molecule.*`. If there's some reason you can't access the Molecule class, this is okay.

File: avogadro/rendering/geometryvisitor.h [hunk 17]
```diff
@@ -62,6 +64,22 @@ class GeometryVisitor : public Visitor
    */
   float radius();
 
+  /**
+   * <<API Extension for TDX>>
```
Comment (ghutchis) [inline avogadro/rendering/geometryvisitor.h:68]: Please, please omit "API Extension." I'm more than happy to provide credit / acknowledgement for your work and TDX, but we don't need to clutter the documentation with who wrote particular methods.

File: avogadro/rendering/scene.cpp [hunk 18]
```diff
@@ -53,4 +53,14 @@ void Scene::clear()
   m_dirty = true;
 }
 
+void Scene::getBoundingBox(double& minX, double& minY, double& minZ,
+                           double& maxX, double& maxY, double& maxZ,
+                           const std::vector<bool>& flags)
+{
+  GeometryVisitor visitor;
```
Comment (ghutchis) [inline avogadro/rendering/scene.cpp:60]: As above - does this need to be in the `Scene` class?

File: avogadro/rendering/scene.h [hunk 19]
```diff
@@ -131,10 +131,26 @@ class AVOGADRORENDERING_EXPORT Scene
   /** Clear the scene of all elements. */
   void clear();
 
+  /**
+   * <<API Extension for TDX>>
```
Comment (ghutchis) [inline avogadro/rendering/scene.h:135]: Please remove this line.

File: avogadro/rendering/scene.h [hunk 20]
```diff
@@ -131,10 +131,26 @@ class AVOGADRORENDERING_EXPORT Scene
...
+   * @param minX [out] minimum X coordinate of the box diagonal
+   * @param minY [out] minimum Y coordinate of the box diagonal
+   * @param minZ [out] minimum Z coordinate of the box diagonal
+   * @param maxX [out] maximum X coordinate of the box diagonal
+   * @param maxY [out] maximum Y coordinate of the box diagonal
+   * @param maxZ [out] maximum Z coordinate of the box diagonal
+   * @param flags [in] flags informing which atoms will be included
+   * in the bounding box
+   */
+  void getBoundingBox(double& minX, double& minY, double& minZ, double& maxX,
```
Comment (ghutchis) [inline avogadro/rendering/scene.h:146]: Again, I think it would be easier to access the bounding box directly from the molecule (which could adjust as needed for isosurface meshes, etc.)

**Review by ghutchis — CHANGES_REQUESTED** [review]
This looks much better, but please avoid using "API Extension" comments, which aren't needed and let's consider putting the bounding box into the `Molecule` class instead.

**Comment by cryos** [comment]
I have been really busy the last month, sorry. I can take a look, you covered a lot of my initial concerns with comments and where the additional API is best placed @ghutchis

File: avogadro/core/molecule.cpp [hunk 21]
```diff
@@ -1324,4 +1324,53 @@ std::list<Index> Molecule::getAtomsAtLayer(size_t layer)
...
 }
 
+void Molecule::boundingBox(Vector3& boxMin, Vector3& boxMax) const
+{
+  boxMin.setConstant(std::numeric_limits<double>::max());
+  boxMax.setConstant(-std::numeric_limits<double>::max());
+
+  bool noSelection = true;
+
+  for (uint32_t i = 0; i < m_selectedAtoms.size(); i++) {
```
Comment (ghutchis) [inline avogadro/core/molecule.cpp:1334]: You can get this from `isSelectionEmpty()` but it's not a big deal either way.

File: avogadro/core/molecule.cpp [hunk 22]
```diff
@@ -1324,4 +1324,53 @@ std::list<Index> Molecule::getAtomsAtLayer(size_t layer)
...
+    if (m_selectedAtoms[i]) {
+      noSelection = false;
+      break;
+    }
+  }
+
+  for (uint32_t i = 0; i < atomCount(); i++) {
+    if (noSelection || m_selectedAtoms[i]) {
+      double radius = 1.0;
```
Comment (ghutchis) [inline avogadro/core/molecule.cpp:1343]: This seems small. I'd suggest adding it as a parameter with a default of 1.0?

File: avogadro/core/molecule.cpp [hunk 23]
```diff
@@ -1324,4 +1324,53 @@ std::list<Index> Molecule::getAtomsAtLayer(size_t layer)
...
+
+      Vector3 boxMinBuffer;
+      Vector3 boxMaxBuffer;
+
+      boxMinBuffer.x() = atom(i).position3d().x() - radius;
```
Comment (ghutchis) [inline avogadro/core/molecule.cpp:1348]: boxMinBuffer = atom(i).position3d().array() - radius;

https://stackoverflow.com/a/35693493/131896

File: avogadro/core/molecule.cpp [hunk 24]
```diff
@@ -1324,4 +1324,53 @@ std::list<Index> Molecule::getAtomsAtLayer(size_t layer)
...
+      boxMinBuffer.y() = atom(i).position3d().y() - radius;
+      boxMinBuffer.z() = atom(i).position3d().z() - radius;
+      boxMaxBuffer.x() = atom(i).position3d().x() + radius;
```
Comment (ghutchis) [inline avogadro/core/molecule.cpp:1351]: boxMaxBuffer = atom(i).position3d().array() + radius;

This way Eigen will properly perform an element-wide operation.

File: avogadro/core/molecule.cpp [hunk 25]
```diff
@@ -1324,4 +1324,53 @@ std::list<Index> Molecule::getAtomsAtLayer(size_t layer)
...
+      boxMaxBuffer.y() = atom(i).position3d().y() + radius;
+      boxMaxBuffer.z() = atom(i).position3d().z() + radius;
+
+      if (boxMinBuffer.x() < boxMin.x())
+        boxMin.x() = boxMinBuffer.x();
```
Comment (ghutchis) [inline avogadro/core/molecule.cpp:1356]: boxMin.x() = min(boxMin.x(), boxMinBuffer.x());

etc. .. which is more readable, since the library handles the conditional

File: avogadro/core/molecule.h [hunk 26]
```diff
@@ -717,9 +718,18 @@ class AVOGADROCORE_EXPORT Molecule
   Layer& layer();
   const Layer& layer() const;
 
+  /**
+   * Calculte and return bounding box of the whole molecule or selected atoms
+   * only.
+   * @param boxMin [out] the minimum corner (first end of the box diagonal)
+   * @param boxMax [out] the maximum corner (second end of the box diagonal)
+   */
+  void boundingBox(Vector3& boxMin, Vector3& boxMax) const;
```
Comment (ghutchis) [inline avogadro/core/molecule.h:727]: Please add  Real radius = 1.0 as an additional parameter

**Review by ghutchis — CHANGES_REQUESTED** [review]
This all looks great .. just some minor nitpicks in the bounding box.

File: avogadro/core/molecule.h (hunk 26 above)
Comment (pskowronskiTDx) [reply avogadro/core/molecule.h]: Understood, suggested changes were introduced.

**Comment by ghutchis** [comment]
Looks good to me - I'll re-run clang-format when I get a chance. Thanks for sticking with this, I'll turn on the flag for Windows and Mac builds.
